nomad/structs/funcs.go

```diff
@@ -187,8 +187,12 @@ func AllocsFit(node *Node, allocs []*Allocation, netIdx *NetworkIndex, checkDevi
 	if netIdx == nil {
 		netIdx = NewNetworkIndex()
 		defer netIdx.Release()
-		if netIdx.SetNode(node) || netIdx.AddAllocs(allocs) {
-			return false, "reserved port collision", used, nil
+
+		if collision, reason := netIdx.SetNode(node); collision {
+			return false, fmt.Sprintf("reserved port collision: %v", reason), used, nil
+		}
+		if collision, reason := netIdx.AddAllocs(allocs); collision {
+			return false, fmt.Sprintf("reserved port collision: %v", reason), used, nil
```

Comment:
Maybe differentiate the errors? AFAIK `SetNode` should *never* collide and *always* indicates a bug (either due to missing configuration validation or incorrect logic).

`AddAllocs` *can* legitimately* collide here if the plan was made concurrently with other allocation changes targeting that node, so there's no reason for users to intervene in this case.

AFAICT the only plan metrics we have are timers and the queue depth gauge. I think knowing the rate of plan application failures would be useful as it always means *something.* If it's consistently non-zero then you should probably lower `num_schedulers` because the concurrent scheduling is conflicting so often it's having to redo work. If it's still non-zero with a low `num_schedulers` then there's likely a bug!

The Plan.Submit RPC might be a reasonable place for this metric as the dequeue/evaluate main loop in plan_apply.go is ...complex.

`*` Where "legitimately" means: can collide even in the absence of bugs.

Reply:
Added new counter metric `nomad.plan.node_rejected` metric 👍